wom/src/main/scala/wom/types/WomType.scala

```diff
@@ -98,4 +98,13 @@ object WomType {
       case t1 if types.forall(t2 => t1.isCoerceableFrom(t2)) => t1
     } getOrElse WomAnyType
   }
+
+  object RecursiveType {
+    def unapply(in: WomType): Option[WomType] = in match {
+      case WomOptionalType(other) => Some(other)
+      case WomMaybeEmptyArrayType(other) => Some(other)
+      case WomNonEmptyArrayType(other) => Some(other)
```

Comment:
Should `WomMapType` and `WomObjectTypeLike` be there too ?

Reply:
Mm I don't think so.  Those are more specific, i.e. WomMapType has two types and we want both of those, and I think a WomObject is not recursive, i.e. we want that whole object and not some piece of it.